File: sample/client/proto/Forum.PostData.cs
// sample/proto/Forum.proto
using ILib.ProtoPack;
using System.Collections.Generic;

namespace Forum
{
	[System.Serializable]
	public partial class PostData : IMessage
	{
		public int Id;

		public string Message;

		public Forum.User User;

		#region Serialization

		/// <summary>
		/// Serialize Message
		/// </summary>
		public void Write(IWriter w)
		{
			// Write Map Length
			w.WriteMapHeader(3);
			
			// Write Id
			w.WriteTag(1);
			w.Write(Id);
			
			// Write Message
			w.WriteTag(2);
			w.Write(Message);
			
			// Write User
			w.WriteTag(3);
			w.Write(User);
		}

		/// <summary>
		/// Deserialize Message
		/// </summary>
		public void Read(IReader r)
		{
			// Read Map Length
			var len = r.ReadMapHeader();

			for (var i = 0; i < len; i++)
			{
				var tag = r.ReadTag();
				switch(tag) {
				case 1: // Read Id
					Id = r.ReadInt();
					break;
				case 2: // Read Message
					Message = r.ReadString();
					break;
				case 3: // Read User
					User = r.ReadMessage<Forum.User>();
					break;
				default:
					r.Skip();
					break;
				}
			}
		}
		#endregion

	}
}

File: sample/goclient/proto/forum.postdata.go
// sample/proto/Forum.proto

package proto

import (
	protopack "github.com/yazawa-ichio/protoc-gen-msgpack/lib/golang"
)

type Forum_PostData struct {
	ID      int32
	Message string
	User    *Forum_User
}

// Pack Serialize Message
func (m *Forum_PostData) Pack() ([]byte, error) {
	return protopack.Pack(m)
}

// Write Serialize Message
func (m *Forum_PostData) Write(w protopack.Writer) error {
	// Write Map Length
	err := w.WriteMapHeader(3)
	if err != nil {
		return err
	}
	
	// Write Id
	err = w.WriteTag(1)
	if err != nil {
		return err
	}
	err = w.WriteInt32(m.ID)
	if err != nil {
		return err
	}
	
	// Write Message
	err = w.WriteTag(2)
	if err != nil {
		return err
	}
	err = w.WriteString(m.Message)
	if err != nil {
		return err
	}
	
	// Write User
	err = w.WriteTag(3)
	if err != nil {
		return err
	}
	err = w.WriteMessage(m.User)
	if err != nil {
		return err
	}
	return err
}

// Unpack Serialize Message
func (m *Forum_PostData) Unpack(buf []byte) (error) {
	return protopack.Unpack(m, buf)
}

// Read Deserialize Message
func (m *Forum_PostData) Read(r protopack.Reader) error {
	// Read Map Length
	len, err := r.ReadMapHeader()
	if err != nil {
		return err
	}

	for i := uint(0); i < len; i++ {

		// Read Tag
		tag, err := r.ReadTag()
		if err != nil {
			return err
		}

		switch tag {
		case 1: // Read Id
			m.ID, err = r.ReadInt32();
			if err != nil {
				return err
			}
			break
		case 2: // Read Message
			m.Message, err = r.ReadString();
			if err != nil {
				return err
			}
			break
		case 3: // Read User
			isNil, err := r.NextFormatIsNull()
			if err != nil {
				return err
			}
			if isNil {
				m.User = nil
				err = r.ReadNil()
			} else {
				m.User = &Forum_User{}
				err = m.User.Read(r)
			}
			if err != nil {
				return err
			}
			break
		default:
			err = r.Skip()
			if err != nil {
				return err
			}
			break
		}
	}
	return err
}
